compiler/rustc_target/src/spec/base/apple/mod.rs

```diff
@@ -97,53 +98,95 @@ impl TargetAbi {
 }
 
 fn pre_link_args(os: &'static str, arch: Arch, abi: TargetAbi) -> LinkArgs {
+    let mut args = LinkArgs::new();
+    // From the man page for ld64 (`man ld`):
+    // > The linker accepts universal (multiple-architecture) input files,
+    // > but always creates a "thin" (single-architecture), standard Mach-O
+    // > output file. The architecture for the output file is specified using
+    // > the -arch option.
+    //
+    // The linker has heuristics to determine the desired architecture, but to
+    // be safe, and to avoid a warning, we set the architecture explicitly.
+    //
+    // Supported architecture names can be found in the source:
+    // https://github.com/apple-oss-distributions/ld64/blob/ld64-951.9/src/abstraction/MachOFileAbstraction.hpp#L578-L648
+    let ld_arch = match arch {
+        Armv7k => "armv7k",
+        Armv7s => "armv7s",
+        Arm64 => "arm64",
+        Arm64e => "arm64e",
+        Arm64_32 => "arm64_32",
+        // ld64 doesn't understand i686, so fall back to i386 instead
+        //
+        // Same story when linking with cc, since that ends up invoking ld64.
+        I386 | I686 => "i386",
+        X86_64 => "x86_64",
+        X86_64h => "x86_64h",
```

Comment:
Maybe factor out this into a function and put it next to `fn target_name` to highlight their similarities and differences?

Reply:
Yeah, kinda wanted to get rid of `target_name` too, and instead just inline it at the usage sites, I felt that was clearer. But anyhow, I've done as you've suggested.